### ISSUE TITLE
undefined reference __tls_get_addr while linking

### ISSUE THREAD

**Issue by lamgea** [ziglang/zig#20625]
### Zig Version

0.13.0

### Steps to Reproduce and Observed Behavior

In the newly created Zig package (created with `zig init`):
1. Referencing the `add` function in `root.zig` from `main.zig`  either by `extern fn` or `@import`.
2. Change the library to be dynamically linked and link the library in the build script.

A compilation error will be thrown when trying to build: 
`ld.lld: undefined reference due to --no-allow-shlib-undefined: __tls_get_addr`

It seem to be specific to Linux. I tried cross compiling to Windows and it worked fine.
Related issue in earlier version: #4748

Additionally, append `.single_threaded = true` to the build option of the library could make it build, but it will give a seg fault when running.

### Expected Behavior

It should build and run without errors.

**Comment by zhylmzr** [comment]
I was unable to reproduce this issue by cross-compiling to Linux:
```
zig1 » zig build -Dtarget=x86_64-linux

zig1 » file zig-out/bin/zig1
zig-out/bin/zig1: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, with debug_info, not stripped

zig1 » zig version
0.13.0
```

```zig
// build.zig
const lib = b.addSharedLibrary(.{
    .name = "zig1",
    .root_source_file = b.path("src/root.zig"),
    .target = target,
    .optimize = optimize,
});

const exe = b.addExecutable(.{
    .name = "zig1",
    .root_source_file = b.path("src/main.zig"),
    .target = target,
    .optimize = optimize,
});

exe.linkLibrary(lib);
```

```zig
// main.zig
const std = @import("std");

extern fn add(a: i32, b: i32) i32;

pub fn main() !void {
    std.debug.print("hello {}", .{add(1, 2)});
}
```

```zig
// root.zig
export fn add(a: i32, b: i32) i32 {
    return a + b;
}
```

**Comment by lamgea** [comment]
@zhylmzr Strange, by explicit about target the binary builds. But it will still run into segfault.
I also reproduced it on Ubuntu image in docker.

**Comment by zhylmzr** [comment]
Yes, although the build will not fail when cross-compiling, but a segfault will occur at runtime.

A temporary solution:

```zig
// build.zig
const lib = b.addSharedLibrary(.{
    .name = "zig1",
    // In this case the main source file is merely a path, however, in more
    // complicated build scripts, this could be a generated file.
    .root_source_file = b.path("src/root.zig"),
    .target = target,
    .optimize = optimize,
    .sanitize_thread = true, // <---- add the line
});
```

**Comment by lamgea** [comment]
@zhylmzr Thanks for that.
I tested it, and it seems like I only need to link libc to get this working.

**Comment by alexrp** [comment]
The fundamental problem here is that `__tls_get_addr` is expected to be present for PIC. It is normally provided by libc, but obviously, if you don't link libc, you don't get an implementation of it.

This should probably go in `std.os.linux.tls`.

**Comment by alexrp** [comment]
Upon further thought, I've decided that this is expected behavior.

`__tls_get_addr` is normally provided by the dynamic linker and its implementation is tightly coupled to the dynamic linker's TLS implementation. This means that we cannot really provide a generic implementation in Zig.

So a user has multiple paths forward here:

* Don't use `--no-allow-shlib-undefined`.
* Link libc (thus linking the dynamic linker).
* Link the dynamic linker explicitly.
* Provide a `__tls_get_addr` implementation.

Which of these is appropriate depends on the use case, of course.
